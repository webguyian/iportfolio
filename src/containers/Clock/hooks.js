import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
