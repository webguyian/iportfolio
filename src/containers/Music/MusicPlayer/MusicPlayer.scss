.music-player {
  @include stretch-column-nowrap;
  align-items: center;
  padding: $gutter-m $gutter $gutter;

  &-track {
    &-cover,
    &-info {
      margin-bottom: $gutter-m;
    }

    &-cover {
      @include flex-center;
      width: 350px;
      height: 350px;
      border-radius: $border-radius-m;
      background-color: $gray-dark;
      color: $gray-medium;
    }

    &-info {
      width: 100%;
    }

    &-name,
    &-artist {
      font-size: $font-size-l;
    }

    &-artist {
      margin-top: $gutter-xs;
      color: $pink;
    }

    &-audio {
      margin-top: auto;
    }
  }

  &-controls {
    display: flex;
    align-items: center;
    justify-content: space-around;
    width: 100%;
    margin: $gutter $gutter $gutter-l;
  }

  &-control {
    @include circle(45px);
    z-index: 2;
    position: relative;
    color: $white;

    &::after {
      @include circle(45px);
      z-index: -1;
      position: absolute;
      transition: transform 0.4s, background-color 0.8s;
      content: "";
    }

    &--play,
    &--play::after {
      width: 55px;
      height: 55px;
    }

    &:hover,
    &:focus {
      &::after {
        transform: scale(1.75);
        background-color: $gray-dark;
      }
    }
  }

  &-volume-control {
    display: flex;
    align-items: center;
    justify-content: space-around;
    width: 100%;

    > .ui-icon {
      color: $gray;
    }
  }

  &-volume-control-slider {
    width: 80%;
    margin: $gutter-s 0;
    background-color: transparent;
    -webkit-appearance: none;

    &:focus {
      outline: none;

      &::-webkit-slider-thumb {
        outline: 2px solid $blue;
        background-color: $gray-light;
      }
    }
  }

  &-volume-control-slider::-webkit-slider-runnable-track {
    width: 100%;
    height: 8.4px;
    border: 0.2px solid $black;
    border-radius: 1.3px;
    background: $gray-dark;
    cursor: pointer;
  }

  &-volume-control-slider::-webkit-slider-thumb {
    width: 20px;
    height: 20px;
    margin-top: -6px;
    transition: color 0.4s;
    border: 2px solid $gray-dark;
    border-radius: 20px;
    background: $gray;
    cursor: pointer;
    -webkit-appearance: none;
  }

  &-volume-control-slider:focus::-webkit-slider-runnable-track {
    background: $gray-dark;
  }

  &-volume-control-slider::-moz-range-track {
    width: 100%;
    height: 8.4px;
    border: 0.2px solid $black;
    border-radius: 1.3px;
    background: $gray;
    cursor: pointer;
  }

  &-volume-control-slider::-moz-range-thumb {
    width: 20px;
    height: 20px;
    transition: color 0.4s;
    border: 2px solid $gray-dark;
    border-radius: 20px;
    background: $gray;
    cursor: pointer;
  }
}
