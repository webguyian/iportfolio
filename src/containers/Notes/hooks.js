import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
