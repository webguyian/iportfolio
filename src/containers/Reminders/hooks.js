import { useEffect, useState, useRef } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useLocalStorage = (key, initialValue) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error return initialValue
      return initialValue;
    }
  };
  const [storedValue, setStoredValue] = useState(getValue);
  const setValue = value => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;

      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      // Drop error
    }
  };

  return [storedValue, setValue, getValue];
};

export const useRefFocus = (node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element) {
      // Add focus to element
      element.focus();
    }
  }, [ref]);

  return ref;
};

export const useRefControlledFocus = (focused, node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element && focused) {
      element.focus();
    }
  }, [focused]);

  return ref;
};

export const useReminders = (reminders, setReminders) => {
  const remindersRef = useRef(reminders);

  useEffect(() => {
    // Update reminders in ref
    remindersRef.current = reminders;
  });

  useEffect(() => {
    return () => {
      const { current } = remindersRef;
      const updatedReminders = current.filter(reminder =>
        reminder.value.trim()
      );

      // Trigger callback without empty reminders
      setReminders(updatedReminders);
    };
  }, [remindersRef]);

  return remindersRef;
};

export const useSwipeOffset = (callback, initial = 0) => {
  const [offset, setOffset] = useState(initial);

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Left') {
      // Exit early if not left swipe
      return false;
    }

    setOffset(eventInfo.deltaX);
  };

  const handleSwipeLeft = eventInfo => {
    const { deltaX } = eventInfo;

    if (deltaX > 80) {
      callback(eventInfo);
    }
  };

  const handleSwipeRight = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedLeft: handleSwipeLeft,
    onSwipedRight: handleSwipeRight,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset, setOffset];
};
