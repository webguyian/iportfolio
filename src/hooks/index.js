import { useEffect, useState } from 'react';

export const useBreakpoint = breakpoint => {
  const [matchesBreakpoint, setBreakpoint] = useState(false);
  const mediaQuery = `(max-width: ${breakpoint}px)`;

  const handleResize = () => {
    const matches = window.matchMedia(mediaQuery).matches;

    setBreakpoint(matches);
  };

  useEffect(() => {
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [breakpoint]);

  return matchesBreakpoint;
};
