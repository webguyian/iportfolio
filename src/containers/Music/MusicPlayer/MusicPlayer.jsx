import React from 'react';
import { object } from 'prop-types';

import Button from 'components/Button/Button';
import Icon from 'components/Icon/Icon';
import Text from 'components/Text/Text';

import { ALBUM_URL } from 'modules/music/constants';
import { usePlayer } from 'modules/music/hooks';

const MusicPlayer = props => {
  const { location } = props;
  const [player, actions] = usePlayer(location);
  const { ref, track } = player;
  const trackName = track ? track.name : 'Not Playing';
  const baseClass = 'music-player';
  const controlClass = `${baseClass}-control`;
  const playClass = `${controlClass} ${controlClass}--play`;
  const trackClass = `${baseClass}-track`;
  const volumeClass = `${baseClass}-volume-control`;

  return (
    <div className={baseClass}>
      <div className={`${trackClass}-cover`}>
        {track ? (
          <img
            className={`${trackClass}-cover-image`}
            alt={`${track.albumName} album cover`}
            src={`${ALBUM_URL}/${track.albumId}/images/350x350.jpg`}
          />
        ) : (
          <Icon name="music" size="10x" />
        )}
      </div>
      <div className={`${trackClass}-info`}>
        <Text className={`${trackClass}-name`} element="h1" modifier="bold">
          {trackName}
        </Text>
        <Text className={`${trackClass}-artist`} element="h2">
          {track && track.artistName}
        </Text>
      </div>
      <audio className={`${trackClass}-audio`} ref={ref}>
        {track && <source src={track.previewURL} type="audio/mpeg" />}
      </audio>
      <div className={`${baseClass}-controls`}>
        <Button
          className={controlClass}
          icon="fast-backward"
          size="3x"
          onClick={actions.prev}
        >
          Previous track
        </Button>
        {player.playing ? (
          <Button
            className={playClass}
            icon="pause"
            size="4x"
            onClick={actions.toggle}
          >
            Pause
          </Button>
        ) : (
          <Button
            className={playClass}
            icon="play"
            size="4x"
            onClick={actions.toggle}
          >
            Play
          </Button>
        )}
        <Button
          className={controlClass}
          icon="fast-forward"
          size="3x"
          onClick={actions.next}
        >
          Next track
        </Button>
      </div>
      <div className={volumeClass}>
        <Icon name="volume-off" label="Volume off" />
        <input
          className={`${volumeClass}-slider`}
          type="range"
          min="0"
          max="1"
          step="0.1"
          onChange={actions.changeVolume}
          value={player.volume}
        />
        <Icon name="volume-up" label="Volume up" />
      </div>
    </div>
  );
};

MusicPlayer.propTypes = {
  location: object.isRequired
};

export default MusicPlayer;
