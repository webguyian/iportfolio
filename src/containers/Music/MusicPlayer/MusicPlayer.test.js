import React from 'react';
import { act, render } from '@testing-library/react';

import * as hooks from 'modules/music/hooks';
import MusicPlayer from './MusicPlayer';

describe('<MusicPlayer />', () => {
  global.HTMLMediaElement.prototype.load = jest.fn();
  global.HTMLMediaElement.prototype.play = jest.fn();
  global.HTMLMediaElement.prototype.pause = jest.fn();

  const usePlayer = hooks.usePlayer;

  hooks.usePlayer = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.usePlayer.mockImplementation(usePlayer);
  });

  const props = {
    location: {}
  };

  const actions = {
    toggle: jest.fn(),
    changeVolume: jest.fn()
  };

  it('renders correctly', async () => {
    let component;

    await act(async () => {
      component = render(<MusicPlayer {...props} />);
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('renders correctly with track', async () => {
    const track = {
      albumId: 'alb.396711596',
      albumName: 'Dance Monkey',
      artistId: 'art.352193935',
      artistName: 'Tones and I',
      disc: 1,
      href: 'https://api.napster.com/v2.2/tracks/tra.396711597',
      id: 'tra.396711597',
      index: 1,
      isrc: 'QZES71982312',
      name: 'Dance Monkey',
      playbackSeconds: 209,
      previewURL: 'https://listen.hs.llnwd.net/g3/9/1/0/8/1/1536018019.mp3',
      shortcut: 'tones-and-i/dance-monkey-single/dance-monkey',
      type: 'track',
      timestamp: 1595086315844
    };

    hooks.usePlayer.mockReturnValueOnce([
      {
        track,
        playing: false,
        ref: jest.fn(),
        volume: 0.2
      },
      actions
    ]);
    const state = { selected: 1 };

    let component;

    await act(async () => {
      component = render(<MusicPlayer location={{ state }} />);
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('renders correctly with track playing', async () => {
    const track = {
      albumId: 'alb.396711596',
      albumName: 'Dance Monkey',
      artistId: 'art.352193935',
      artistName: 'Tones and I',
      name: 'Dance Monkey',
      previewURL: 'https://listen.hs.llnwd.net/g3/9/1/0/8/1/1536018019.mp3'
    };

    hooks.usePlayer.mockReturnValueOnce([
      {
        track,
        playing: true,
        ref: jest.fn(),
        volume: 0.2
      },
      actions
    ]);
    const state = { selected: 1 };

    let component;

    await act(async () => {
      component = render(<MusicPlayer location={{ state }} />);
    });

    expect(component.asFragment()).toMatchSnapshot();
  });
});
