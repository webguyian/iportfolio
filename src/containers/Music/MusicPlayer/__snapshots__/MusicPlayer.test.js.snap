// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MusicPlayer /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="music-player"
  >
    <div
      class="music-player-track-cover"
    >
      <i
        class="ui-icon ui-icon-music fas fa-music fa-10x"
      />
    </div>
    <div
      class="music-player-track-info"
    >
      <h1
        class="ui-text ui-text--body ui-text--bold music-player-track-name"
      >
        Not Playing
      </h1>
      <h2
        class="ui-text ui-text--body music-player-track-artist"
      />
    </div>
    <audio
      class="music-player-track-audio"
    />
    <div
      class="music-player-controls"
    >
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-backward fas fa-fast-backward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Previous track
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control music-player-control--play"
        type="button"
      >
        <i
          class="ui-icon ui-icon-play fas fa-play fa-4x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Play
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-forward fas fa-fast-forward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Next track
        </span>
      </button>
    </div>
    <div
      class="music-player-volume-control"
    >
      <i
        class="ui-icon ui-icon-volume-off fas fa-volume-off"
        title="Volume off"
      />
      <input
        class="music-player-volume-control-slider"
        max="1"
        min="0"
        step="0.1"
        type="range"
        value="0.2"
      />
      <i
        class="ui-icon ui-icon-volume-up fas fa-volume-up"
        title="Volume up"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MusicPlayer /> renders correctly with track 1`] = `
<DocumentFragment>
  <div
    class="music-player"
  >
    <div
      class="music-player-track-cover"
    >
      <img
        alt="Dance Monkey album cover"
        class="music-player-track-cover-image"
        src="https://direct.rhapsody.com/imageserver/v2/albums/alb.396711596/images/350x350.jpg"
      />
    </div>
    <div
      class="music-player-track-info"
    >
      <h1
        class="ui-text ui-text--body ui-text--bold music-player-track-name"
      >
        Dance Monkey
      </h1>
      <h2
        class="ui-text ui-text--body music-player-track-artist"
      >
        Tones and I
      </h2>
    </div>
    <audio
      class="music-player-track-audio"
    >
      <source
        src="https://listen.hs.llnwd.net/g3/9/1/0/8/1/1536018019.mp3"
        type="audio/mpeg"
      />
    </audio>
    <div
      class="music-player-controls"
    >
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-backward fas fa-fast-backward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Previous track
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control music-player-control--play"
        type="button"
      >
        <i
          class="ui-icon ui-icon-play fas fa-play fa-4x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Play
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-forward fas fa-fast-forward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Next track
        </span>
      </button>
    </div>
    <div
      class="music-player-volume-control"
    >
      <i
        class="ui-icon ui-icon-volume-off fas fa-volume-off"
        title="Volume off"
      />
      <input
        class="music-player-volume-control-slider"
        max="1"
        min="0"
        step="0.1"
        type="range"
        value="0.2"
      />
      <i
        class="ui-icon ui-icon-volume-up fas fa-volume-up"
        title="Volume up"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MusicPlayer /> renders correctly with track playing 1`] = `
<DocumentFragment>
  <div
    class="music-player"
  >
    <div
      class="music-player-track-cover"
    >
      <img
        alt="Dance Monkey album cover"
        class="music-player-track-cover-image"
        src="https://direct.rhapsody.com/imageserver/v2/albums/alb.396711596/images/350x350.jpg"
      />
    </div>
    <div
      class="music-player-track-info"
    >
      <h1
        class="ui-text ui-text--body ui-text--bold music-player-track-name"
      >
        Dance Monkey
      </h1>
      <h2
        class="ui-text ui-text--body music-player-track-artist"
      >
        Tones and I
      </h2>
    </div>
    <audio
      class="music-player-track-audio"
    >
      <source
        src="https://listen.hs.llnwd.net/g3/9/1/0/8/1/1536018019.mp3"
        type="audio/mpeg"
      />
    </audio>
    <div
      class="music-player-controls"
    >
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-backward fas fa-fast-backward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Previous track
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control music-player-control--play"
        type="button"
      >
        <i
          class="ui-icon ui-icon-pause fas fa-pause fa-4x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Pause
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon music-player-control"
        type="button"
      >
        <i
          class="ui-icon ui-icon-fast-forward fas fa-fast-forward fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Next track
        </span>
      </button>
    </div>
    <div
      class="music-player-volume-control"
    >
      <i
        class="ui-icon ui-icon-volume-off fas fa-volume-off"
        title="Volume off"
      />
      <input
        class="music-player-volume-control-slider"
        max="1"
        min="0"
        step="0.1"
        type="range"
        value="0.2"
      />
      <i
        class="ui-icon ui-icon-volume-up fas fa-volume-up"
        title="Volume up"
      />
    </div>
  </div>
</DocumentFragment>
`;
